text/0047-random-assignment-of-availability-chunks.md

```diff
@@ -0,0 +1,179 @@
+1. The `pov_recovery` task of the collators starts availability recovery in response to a `CandidateBacked` event, which enables
+easy access to the core index the candidate started occupying.
+1. Disputes may be initiated on a number of occasions:
+  
+    3.a. is initiated by the current node as a result of finding an invalid candidate while doing approval work. In this case,
+  availability-recovery is not needed, since the validator already issued their vote.
+    
+    3.b is initiated by noticing dispute votes recorded as a result of chain scraping. In this case, the subsystem
+  assumes it already has a copy of the `CandidateReceipt` as a result of scraping `CandidateBacked` events. The logic can be modified
+  to also record core indices.
+  
+    3.c is initiated as a result of getting a dispute statement from another validator. It is possible that the dispute is happening on
+  a fork that was not yet imported by this validator, so the subsystem will not always have access to the `CandidateBacked` event
+  recorded by the chain scraper.
+
+As a solution to 3.c, a new version for the disputes request-response networking protocol will be added.
+This message type will include the relay block hash where the candidate was included. This information will be used
+in order to query the runtime API and retrieve the core index that the candidate was occupying.
+
+The usage of the systematic data availability recovery feature will also be subject to all nodes using the V2 disputes networking protocol.
+
+#### Alternatives to using core_index
+
+As an alternative to core_index, the `ParaId` could be used. It has the advantage of being readily available in the
+`CandidateReceipt`, which would enable the dispute communication protocol to not change and would simplify the implementation.
+However, in the context of [CoreJam](https://github.com/polkadot-fellows/RFCs/pull/31), `ParaId`s will no longer exist (at least not in their current form).
+
+## Drawbacks
+
+Has a fair amount of technical complexity involved:
+
+- Introduces another runtime API that is going to be issued by multiple subsystems. With adequate client-side caching, this should be acceptable.
+
+- Requires a networking protocol upgrade on the disputes request-response protocol
+
+## Testing, Security, and Privacy
+
+Extensive testing will be conducted - both automated and manual.
+This proposal doesn't affect security or privacy.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+This is a necessary optimisation, as reed-solomon erasure coding has proven to be a top consumer of CPU time in polkadot when scaling the parachain count.
+
+With this optimisation, preliminary performance results show that CPU time used for reed-solomon coding can be halved and total POV recovery time decrease by 80% for large POVs. See more [here](https://github.com/paritytech/polkadot-sdk/issues/598#issuecomment-1792007099).
```

Comment:
Calling into the runtime to get chunk mappings is slower than doing that in the client, especially doing so many calls and building up overhead. I suggest only a bulk API, so it's just one call per candidate.

Reply:
There are places (like bitfield signing) that don't need to query the chunk indices of all the validators, since we only need to attest that we have our chunk.

That's why I think it's valuable to have both a bulk API and one for retrieving a single index